doc/user_guide/psyir.rst

```diff
@@ -45,6 +46,56 @@ source code using a frontend. Transformations act on the PSyIR and
 ultimately the generated code is produced by one of the PSyIR's
 backends.
 
+PSyIR Nodes
+-----------
+
+The PSyIR consists of classes whose instances can be connected
+together to form a tree which represent computation in a
+language-independent way. These classes all inherit from the ``Node``
+baseclass and, as a result, PSyIR instances are often referred to
+collectively as 'PSyIR nodes'.
+
+At the present time PSyIR classes can be essentially split into two
+types. PSy-layer classes and Kernel-layer classes. PSy-layer classes
+make use of a ``gen_code()`` or an ``update()`` method to create
+Fortran code whereas Kernel-layer classes make use of PSyIR backends
+to create code.
+
+.. note:: This separation will be removed in the future and eventually
+	  all PSyIR classes will make use of backends with the
+	  expectation that ``gen_code()`` and ``update()`` methods will
+	  be removed.
+
+PSy-layer nodes
+^^^^^^^^^^^^^^^
+
+PSy-layer PSyIR classes are primarily used to create the
+PSy-layer. These tend to be relatively descriptive and do not specify
+how a particular PSyclone frontend would implement them. With the
+exception of ``Loop``, these classes are currently not compatible with
+the PSyIR backends. The generic (non-api-specific) PSyIR nodes are:
+``Schedule``, ``Directive``, ``GlobalSum``, ``HaloExchange``, ``Loop``
+and ``Kern``. The ``Schedule`` class is further subclassed into
+``InvokeSchedule``. The ``Directive`` class is subclassed into many
+directives associated with OpenMP and OpenACC. The ``Kern`` class is
+subclassed into ``CodedKern``, ``InlinedKern`` and ``BuiltinKern``.
+
+
+Kernel-layer nodes
+^^^^^^^^^^^^^^^^^^
+
+Kernel-layer PSyIR classes are currently used to describe existing
+code in a language independent way. Consequently these nodes are more
+prescriptive and are independent of a particular PSyclone
+frontend. These nodes are designed to be used with PSyIR backends. One
+PSy-layer class (``Loop``), can also be used as a Kernel-layer
+class. Kernel-layer PSyIR nodes are: ``Loop``, ``IfBlock``,
+``CodeBlock``, ``Assignment``, ``Reference``, ``Operation``,
+``Literal``, ``Return`` and ``Container``. The ``Operation`` class is
+further subclassed into ``UnaryOperation``, ``BinaryOperation`` and
+``NaryOperation``.
+
```

Comment:
The only issue I have with this classification is regarding ``Schedule``, I think it could be mentioned in the classes that can be used in both layers like the ``Loop`` node and you are missing its subclass ``KernelSchedule`` which is a Kernel-layer node.

Reply:
Also `Array` is missing

Reply:
Very good points. I've updated the documentation and now think that I will need to add create methods for Array and KernelSchedule.